zwrap/sample.go

```diff
@@ -64,36 +64,44 @@ func (c *counters) Reset(key string) {
 // Sample returns a sampling logger. The logger maintains a separate bucket
 // for each message (e.g., "foo" in logger.Warn("foo")). In each tick, the
 // sampler will emit the first N logs in each bucket and every Mth log
-// therafter. Sampling loggers are safe for concurrent use.
+// thereafter (rounded down to nearest power of 2). Sampling loggers are safe for concurrent use.
 //
 // Per-message counts are shared between parent and child loggers, which allows
 // applications to more easily control global I/O load.
 func Sample(zl zap.Logger, tick time.Duration, first, thereafter int) zap.Logger {
+	power := 1
+	for {
+		thereafter >>= 1
+		if thereafter <= 0 {
+			break
+		}
+		power <<= 1
+	}
 	return &sampler{
-		Logger:     zl,
-		tick:       tick,
-		counts:     &counters{counts: make(map[string]*atomic.Uint64)},
-		first:      uint64(first),
-		thereafter: uint64(thereafter),
+		Logger: zl,
+		tick:   tick,
+		counts: &counters{counts: make(map[string]*atomic.Uint64)},
+		first:  uint64(first),
+		power:  uint64(power - 1),
```

Comment:
Using `zl` - I'd like to make sure that the message ends up in the user's application logs.